### PR TITLE
Fslock optimize

### PR DESCRIPTION
**Pull request by crafcat7**
## Summary
1. Add lib_get_pathbuffer / lib_put_pathbuffer instead of temporary variable.
2. Reduce stack size. Getting the tmp buffer through the lib_get_pathbuffer function reduces the declaration of temporary variables
3. Fix an unsynchronized lock acquisition by closelk in multi-threaded scenarios. Ensure atomic behavior when acquiring lock status.
4. Optimized additional performance overhead due to closelk in the close phase (even if the fd doesn't use fslock)
Optimizing performance:
Before: Time taken to close the file: 33984 nsec
After: Time taken to close the file: 23744 nsec
Improvement of about 10 msec

## Impact
Optimizing fslock

## Testing
Local test pass

### CODE REVIEW

**Review by PetervdPerk-NXP — CHANGES_REQUESTED** [review]
You're trading stack usage for heap fragmentation.
I disagree with this approach because once your heap is to fragmented `malloc` would simply fail and your system most likely crashes.
 if you really want it I propose you make an optional kconfig symbol called i.e. `dynamic path allocation`

**Comment by crafcat7** [comment]
> You're trading stack usage for heap fragmentation. I disagree with this approach because once your heap is to fragmented `malloc` would simply fail and your system most likely crashes. if you really want it I propose you make an optional kconfig symbol called i.e. `dynamic path allocation`

lib_get_pathbuffer will initially request an address from a global stack, just like we use temporary variables, which will not cause memory fragmentation. Unless the current usage peak exceeds the available buffer, it will request from the heap.

I think the fragmentation problem can be tuned with CONFIG_LIBC_MAX_PATHBUFFER, which can reduce/avoid malloc from heap

**Comment by guohao15** [comment]
> You're trading stack usage for heap fragmentation. I disagree with this approach because once your heap is to fragmented `malloc` would simply fail and your system most likely crashes. if you really want it I propose you make an optional kconfig symbol called i.e. `dynamic path allocation`

We can use LIBC_MAX_PATHBUFFER Kconfig  to optimise the useasge of this buffer, reduce the chance  that  system alloc memory form heap .

**Comment by PetervdPerk-NXP** [comment]
> if you want, we can add an option to skip the allocation and return NULL directly if the usage exceeds CONFIG_LIBC_MAX_PATHBUFFER.

Yeah that sounds more logical to me. I don't like the idea hidden malloc's.

**Comment by xiaoxiang781216** [comment]
> > if you want, we can add an option to skip the allocation and return NULL directly if the usage exceeds CONFIG_LIBC_MAX_PATHBUFFER.
> 
> Yeah that sounds more logical to me. I don't like the idea of hidden malloc's.

OK, @crafcat7 will create a new patch to forbid the allocation.

**Comment by crafcat7** [comment]
> > > if you want, we can add an option to skip the allocation and return NULL directly if the usage exceeds CONFIG_LIBC_MAX_PATHBUFFER.
> > 
> > 
> > Yeah that sounds more logical to me. I don't like the idea of hidden malloc's.
> 
> OK, @crafcat7 will create a new patch to forbid the allocation.

Done

**Review by PetervdPerk-NXP — APPROVED** [review]

File: fs/vfs/fs_lock.c [hunk 1]
```diff
@@ -602,6 +608,8 @@ int file_getlk(FAR struct file *filep, FAR struct flock *flock)
       flock->l_len = flock->l_end - flock->l_start + 1;
     }
 
+out_free:
+  lib_put_pathbuffer(path);
   return OK;
```
Comment (pkarashchenko) [inline fs/vfs/fs_lock.c:613]: should we `return ret;` here instead?

File: fs/vfs/fs_lock.c [hunk 2]
```diff
@@ -731,36 +752,49 @@ void file_closelk(FAR struct file *filep)
   FAR struct file_lock_bucket_s *bucket;
   FAR struct file_lock_s *file_lock;
   FAR struct file_lock_s *temp;
-  char path[PATH_MAX];
+  FAR char *path;
   bool deleted = false;
   int ret;
 
+  if (filep->locked == false)
```
Comment (pkarashchenko) [inline fs/vfs/fs_lock.c:759]:
```suggestion
  if (!filep->locked)
```

File: fs/vfs/fs_lock.c [hunk 3]
```diff
@@ -774,6 +808,8 @@ void file_closelk(FAR struct file *filep)
     }
 
   nxmutex_unlock(&g_protect_lock);
+out:
+  lib_put_pathbuffer(path);
```
Comment (pkarashchenko) [inline fs/vfs/fs_lock.c:812]: Maybe better
```suggestion
out:
  nxmutex_unlock(&g_protect_lock);
out_free:
  lib_put_pathbuffer(path);
```
and remove `nxmutex_unlock(&g_protect_lock);` before `goto out;`?

File: libs/libc/misc/lib_pathbuffer.c [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct pathbuffer_s
+{
+  mutex_t lock;         /* Lock for the buffer */
+  int free_bitmap;      /* Bitmap of free buffer */
```
Comment (pkarashchenko) [inline libs/libc/misc/lib_pathbuffer.c:43]:
```suggestion
  unsigned int free_bitmap;      /* Bitmap of free buffer */
```

**Comment by xiaoxiang781216** [comment]
@crafcat7 please fix the conflict.

File: libs/libc/misc/Kconfig [hunk 5]
```diff
@@ -120,3 +120,9 @@ config LIBC_MAX_PATHBUFFER
 	---help---
 		This value is the maximum size of the buffer that will hold the full
 		file path.
+
+config LIBC_PATHBUFFER_MALLOC
+	bool "Enable malloc pathbuffer"
+	default n
```
Comment (xiaoxiang781216) [inline libs/libc/misc/Kconfig:126]: n->y?

Comment (crafcat7) [reply libs/libc/misc/Kconfig]: Done

File: fs/vfs/fs_lock.c (hunk 3 above)
Comment (crafcat7) [reply fs/vfs/fs_lock.c]: Done

File: libs/libc/misc/lib_pathbuffer.c (hunk 4 above)
Comment (crafcat7) [reply libs/libc/misc/lib_pathbuffer.c]: Done

File: fs/vfs/fs_lock.c (hunk 2 above)
Comment (crafcat7) [reply fs/vfs/fs_lock.c]: Done

File: fs/vfs/fs_lock.c (hunk 1 above)
Comment (crafcat7) [reply fs/vfs/fs_lock.c]: Done

**Comment by crafcat7** [comment]
> @crafcat7 please fix the conflict.

Done

File: libs/libc/misc/Kconfig [hunk 6]
```diff
@@ -112,3 +112,17 @@ config LIBC_MEM_FD_VFS_PATH
 	depends on !LIBC_MEMFD_ERROR
 	---help---
 		The relative path to where memfd will exist in the tmpfs namespace.
+
+config LIBC_MAX_PATHBUFFER
+	int "Maximum size of a temporary file path buffer array"
+	range 0 32
+	default 2
```
Comment (PetervdPerk-NXP) [inline libs/libc/misc/Kconfig:119]: So this should either be 0 or only be set if fs_lock is enabled.
Otherwise we're wasting 512 bytes of RAM.

Comment (xiaoxiang781216) [reply libs/libc/misc/Kconfig:119]: actually, we don't need do this type of manual optimization, the linker is enough smart to remove the unused code and data from the final image automatically if nobody calls lib_[get|put]_pathbuffer.

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**